src/audio/src/src_generic.c

```diff
@@ -203,16 +193,16 @@ static inline void fir_filter_generic(int32_t *rp, const void *cp, int32_t *wp0,
 		 * Qx.31.
 		 */
 		for (i = 0; i < n1; i += nch) {
-			y0 += (int64_t)(*coef >> 8) * (*data);
-			coef++;
+			y0 += (int64_t)(*(coef++) >> 8) * (*data);
```

Comment:
I am not sure whether we can calculate based on sample by sample, seems current we are calculating based on channel, this will introduce more data cache miss, you can image that, better to handle them sample by sample, instead of channel by channel.

Reply:
@btian1 @lyakh Would it be best if I split this generic C FIR rewrite to own PR and do it later when I have more time available for it? The hifi3 version is now what matters and if any issue there I will fix it immediately. This is much lower priority. The generic code is not broken, it's only slow.

Reply:
I am fine with this.

Reply:
> > @singalsu let me explain a bit why I stumbled over this PR. It proposes to use operations like
> > ```
> > y0 += (int64_t)(*(coef++)) * (*data);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > where it's relying on the `coef` increment being done "after `coef` got used." But `coef++` is enclosed in parentheses which (roughly) means - evaluate expression in parentheses before the rest. In that sense one could (falsely) assume that `coef` is incremented before dereference. I checked my favourite first reference K&R and didn't find a definite answer to that there. A quick online search didn't convince me either way either. So, to me it looks confusing, but it seems to do the right thing...
> 
> The standard states that the increment happens after the read of "coef" in the "coef++" expression and before the next sequence point, which is the ";" at the end of the line. LLVM or GCC must ensure that any reads of "coef" after that ";" will read the new value and thus any further delaying of coef being incremented must be at most before the next read.
> 
> The expression is fine (no UB) but hard to read.
> 
> The standard states nothing about parentheses, other than as something that allows you to bypass the normal associativity rules and to group things. There's no sequence points or anything in there, just grouping. The "evaluate expression in parentheses before the rest" thing is strictly not true, you just have the parentheses to avoid the order of operations being funny (I don't know how unary * and postfix ++ deal with each other, I believe removing the internal parenthesis still works due to the priority of operations in C but it's even more unreadable that way).
> 
> The only operator inside an expression that does introduce sequence points is the comma operator. That is, the expression "x++,x" is valid as the read after the comma is guaranteed to be ordered after the increment before the comma.
> 
> Anyway, my conclusion here is that code looks funny but technically breaks no rules (other than, again, readability which I'll allow you to debate)
> 
> Edit: Also on `In that sense one could (falsely) assume that coef is incremented before dereference.` -- well maybe the increment does happen before the dereference, but you're getting the value from before the increment to be used further on in the expression anyway so does it matter?

I would suggest remove (), like below, it is more clear to me:
y0 += (int64_t)(*coef++) * (*data);
() brings understanding trouble, developer need think the priority between * and ++， without(), at least, * always the highest priority.

Reply:
> I would suggest remove (), like below, it is more clear to me: y0 += (int64_t)(*coef++) * (*data); () brings understanding trouble, developer need think the priority between * and ++， without(), at least, * always the highest priority.

Actually `*` (dereference) and `++` have the same precedence, it's only the associativity that decides on the order here.